_posts/2022-XX-XX-fido2-central-auth.md

```diff
@@ -0,0 +1,125 @@
+---
+layout:     post
+title:      "Local authentication in a centralized management system with FIDO2"
+date:       2022-09-08 12:00:00 +0100
+categories: idm
+---
+
+#TODO: change date
+
+#TODO: change for fido2 helper process for sssctl
+
+# Introduction
+
+I have been working on a feature that will allow centrally managed users to log in locally to a computer with FIDO2 keys. As the feature is still under development I've prepared an environment to test it.
+
+## Dependencies
+
+The test environment uses docker/podman to set up the containers and ansible to provision them. On top of that, yubikey-manager is needed to manage the FIDO2 device options. Moreover, I've created a [COPR repository](https://copr.fedorainfracloud.org/coprs/ipedrosa/fido2-central-auth/) to deliver the patched versions of SSSD and IPA that will enable this feature. Finally, I've prepared a new [sssd-ci-container branch](https://github.com/ikerexxe/sssd-ci-containers/tree/fido2) to automate the provision of everything. The following sections will explain how to set up the environment and how to test it.
+
+# How to set up the environment
+
+1. Clone sssd-ci-containers locally and checkout the fido2 branch:
+
+```console
+$ git clone https://github.com/ikerexxe/sssd-ci-containers/
+$ cd sssd-ci-containers
+$ git checkout --track origin/fido2
+```
+
+2. Host package installation and configuration. The following commands should be executed only once:
+
+```console
+$ sudo dnf install -y podman podman-docker docker-compose yubikey-manager
+$ sudo systemctl enable --now podman.socket
+$ sudo setsebool -P container_manage_cgroup true
+$ cp env.example .env
+```
+
+3. Set up the containers (make up) and update the packages with the FIDO2 patches (make fido2). For the latter command to run correctly you will need to input the password `Secret123`. It will take some time to execute as it needs to update the dnf cache and then update the sssd packages.
+
+```console
+$ sudo make up
+$ sudo make fido2

```

Comment:
Latest error is:
~~~
[justin@agalloch sssd-ci-containers]$ sudo make fido2
ansible-playbook -i src/ansible/fido2/inventory.yml \
--user root --ask-pass src/ansible/fido2/playbook.yml
SSH password: 
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details
ERROR! couldn't resolve module/action 'community.general.copr'. This often indicates a misspelling, missing collection, or incorrect module path.

The error appears to be in '/home/justin/github/sssd-ci-containers/src/ansible/fido2/playbook.yml': line 8, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

  tasks:
    - name: Enable testing repo
      ^ here
make: *** [Makefile:31: fido2] Error 4
~~~

Reply:
Can you check if the community.general module is installed in ansible?

```
ansible-galaxy collection list | grep community.general
```

If installed the previous command should print:
```
community.general             4.8.2
```

If it isn't then install it:
```
ansible-galaxy collection install community.general
```

Reply:
It was not installed but after installing it fails the same way. I'll investigate further later.

~~~
[justin@agalloch sssd-ci-containers]$ sudo ansible-galaxy collection install community.general
Process install dependency map
Starting collection install process
Skipping 'community.general' as it is already installed
[justin@agalloch sssd-ci-containers]$ ansible-galaxy collection install community.general
Process install dependency map
Starting collection install process
Skipping 'community.general' as it is already installed
[justin@agalloch sssd-ci-containers]$ sudo make fido2
ansible-playbook -i src/ansible/fido2/inventory.yml \
--user root --ask-pass src/ansible/fido2/playbook.yml
SSH password: 
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details
ERROR! couldn't resolve module/action 'community.general.copr'. This often indicates a misspelling, missing collection, or incorrect module path.

The error appears to be in '/home/justin/github/sssd-ci-containers/src/ansible/fido2/playbook.yml': line 8, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

  tasks:
    - name: Enable testing repo
      ^ here
make: *** [Makefile:31: fido2] Error 4
~~~

Reply:
You can try with `dnf -y install ansible-collection-community-general` but the preferred solution is to install it with ansible-galaxy.

Reply:
@ikerexxe same issue after installing the rpm. I have been trying on Fedora 35, could you try with `1minutetip 1MT-Fedora-35` and see if you can figure out what is wrong? I tried some basic google searching but could not figure out how to resolve it.

Reply:
I tried it and it seems to work correctly :(

Can you modify the `Makefile` and add `-vvvv` to the fido2 target to print additional logs?

Reply:
> Can you check if the community.general module is installed in ansible?
> 
> ```
> ansible-galaxy collection list | grep community.general
> ```

The `list` command is not available in F35.

~~~
# ansible-galaxy collection list
usage: ansible-galaxy collection [-h] COLLECTION_ACTION ...
ansible-galaxy collection: error: argument COLLECTION_ACTION: invalid choice: 'list' (choose from 'init', 'build', 'publish', 'install')
~~~

I suspect the community general ansible collection version in F35 does not contain the copr module. In rawhide make fido2 is working fine but I had to install the 'sshpass' rpm first.

Reply:
> The `list` command is not available in F35.

I didn't test this command. I just executed the steps provided in the blog post and apart from installing `sshpass` everything worked correctly.

> I suspect the community general ansible collection version in F35 does not contain the copr module.

That might be possible. I'm using fedora 36 and it is available there.